coderd/notifications/reports/generator.go

```diff
@@ -0,0 +1,296 @@
+				map[string]string{
+					"template_name":         stats.TemplateName,
+					"template_display_name": templateDisplayName,
+				},
+				reportData,
+				"report_generator",
+				stats.TemplateID, stats.TemplateOrganizationID,
+			); err != nil {
+				logger.Warn(ctx, "failed to send a report with failed workspace builds", slog.Error(err))
+			}
+		}
+	}
+
+	for recipient := range reportGeneratedNow {
+		err = db.UpsertReportGeneratorLog(ctx, database.UpsertReportGeneratorLogParams{
+			UserID:                 recipient,
+			NotificationTemplateID: notifications.TemplateWorkspaceBuildsFailedReport,
+			LastGeneratedAt:        dbtime.Time(now).UTC(),
+		})
+		if err != nil {
+			logger.Error(ctx, "unable to update report generator logs", slog.F("user_id", recipient), slog.Error(err))
+		}
+	}
+
+	err = db.DeleteOldReportGeneratorLogs(ctx, database.DeleteOldReportGeneratorLogsParams{
+		NotificationTemplateID: notifications.TemplateWorkspaceBuildsFailedReport,
+		Before:                 dbtime.Time(now.Add(-failedWorkspaceBuildsReportFrequencyDays*24*time.Hour - time.Hour)).UTC(),
+	})
+	if err != nil {
+		return xerrors.Errorf("unable to delete old report generator logs: %w", err)
+	}
+	return nil
+}
+
+func buildDataForReportFailedWorkspaceBuilds(frequencyDays int, stats database.GetWorkspaceBuildStatsByTemplatesRow, failedBuilds []database.GetFailedWorkspaceBuildsByTemplateIDRow) map[string]any {
+	// Format frequency label
+	var frequencyLabel string
+	if frequencyDays == 7 {
+		frequencyLabel = "week"
+	} else {
+		var plural string
+		if frequencyDays > 1 {
+			plural = "s"
+		}
+		frequencyLabel = fmt.Sprintf("%d day%s", frequencyDays, plural)
+	}
+
+	// Sorting order: template_version_name ASC, workspace build number DESC
+	sort.Slice(failedBuilds, func(i, j int) bool {
+		if failedBuilds[i].TemplateVersionName != failedBuilds[j].TemplateVersionName {
+			return failedBuilds[i].TemplateVersionName < failedBuilds[j].TemplateVersionName
+		}
+		return failedBuilds[i].WorkspaceBuildNumber > failedBuilds[j].WorkspaceBuildNumber
+	})
+
+	// Build notification model for template versions and failed workspace builds
+	templateVersions := []map[string]any{}
+	for _, failedBuild := range failedBuilds {
+		c := len(templateVersions)
+
+		if len(templateVersions) == 0 || templateVersions[c-1]["template_version_name"] != failedBuild.TemplateVersionName {
+			templateVersions = append(templateVersions, map[string]any{
+				"template_version_name": failedBuild.TemplateVersionName,
+				"failed_count":          1,
+				"failed_builds": []map[string]any{
+					{
+						"workspace_owner_username": failedBuild.WorkspaceOwnerUsername,
+						"workspace_name":           failedBuild.WorkspaceName,
+						"build_number":             failedBuild.WorkspaceBuildNumber,
+					},
+				},
+			})
+			continue
+		}
+
+		//nolint:errorlint,forcetypeassert // only this function prepares the notification model
```

Comment:
Some comments here would be appreciated; it's difficult to understand what you're trying to do here.

Reply:
I will drop detailed comments.

The idea is to review workspace builds, and link them with matching template versions.